src/librekey/key_store_g10.c

```diff
@@ -779,14 +1121,299 @@ write_seckey(s_exp_t *s_exp, pgp_seckey_t *key)
+
+    default:
+        fprintf(stderr, "Unsupported block cipher mode: %d\n", key->cipher_mode);
+        return false;
+    }
+
+    if (!add_sub_sexp_to_sexp(&raw_s_exp, &sub_s_exp)) {
+        destroy_s_exp(&raw_s_exp);
+        return false;
+    }
+
+    if (!write_seckey(sub_s_exp, key)) {
+        destroy_s_exp(&raw_s_exp);
+        return false;
+    }
+
+    if (!write_sexp(&raw_s_exp, &raw)) {
+        destroy_s_exp(&raw_s_exp);
+        pgp_memory_release(&raw);
+        return false;
+    }
+
+    keysize = pgp_key_size(key->alg);
+    if (keysize == 0) {
+        (void) fprintf(stderr, "parse_seckey: unknown symmetric algo");
+        destroy_s_exp(&raw_s_exp);
+        pgp_memory_release(&raw);
+        return false;
+    }
+
+    if (pgp_s2k_iterated(key->hash_alg,
+                         derived_key,
+                         keysize,
+                         (const char *) passphrase,
+                         key->salt,
+                         key->s2k_iterations)) {
+        (void) fprintf(stderr, "pgp_s2k_iterated failed\n");
+        destroy_s_exp(&raw_s_exp);
+        pgp_memory_release(&raw);
+        return false;
+    }
+
+    if (rnp_get_debug(__FILE__)) {
+        hexdump(stderr, "input iv", key->iv, G10_CBC_IV_SIZE);
+        hexdump(stderr, "key", derived_key, keysize);
+    }
+
+    // add padding!
+    key->encrypted_len = 16 + raw.length;
+
+    key->encrypted = malloc(key->encrypted_len);
+    if (key->encrypted == NULL) {
+        (void) fprintf(stderr, "can't allocate memory\n");
+        destroy_s_exp(&raw_s_exp);
+        pgp_memory_release(&raw);
+        return false;
+    }
+
+    switch (key->cipher_mode) {
+    case PGP_CIPHER_MODE_CBC:
+        if (botan_cipher_init(&encrypt, "AES-128/CBC", BOTAN_CIPHER_INIT_FLAG_ENCRYPT)) {
+            (void) fprintf(stderr, "botan_cipher_init failed\n");
+            free(key->encrypted);
+            key->encrypted = NULL;
+            key->encrypted_len = 0;
+            destroy_s_exp(&raw_s_exp);
+            pgp_memory_release(&raw);
+            return false;
+        }
+
+        if (botan_cipher_set_key(encrypt, derived_key, keysize)) {
+            free(key->encrypted);
+            key->encrypted = NULL;
+            key->encrypted_len = 0;
+            destroy_s_exp(&raw_s_exp);
+            pgp_memory_release(&raw);
+            botan_cipher_destroy(encrypt);
+            return false;
+        }
+
+        if (botan_cipher_start(encrypt, key->iv, G10_CBC_IV_SIZE)) {
+            free(key->encrypted);
+            key->encrypted = NULL;
+            key->encrypted_len = 0;
+            destroy_s_exp(&raw_s_exp);
+            pgp_memory_release(&raw);
+            botan_cipher_destroy(encrypt);
+            return false;
+        }
+
+        break;
+
+    case PGP_CIPHER_MODE_OCB:
+        if (botan_cipher_init(&encrypt, "AES-128/OCB", BOTAN_CIPHER_INIT_FLAG_ENCRYPT)) {
+            (void) fprintf(stderr, "botan_cipher_init failed\n");
```

Comment:
a lot of copy-pasta here. Isn't it better to "goto error" end handle this all in one place?